keps/sig-network/20191205-kube-proxy-IP-node-binding.md

```diff
@@ -0,0 +1,81 @@
+
+1. To redirect pod traffic that is intended for LoadBalancer IP directly to the service, bypassing the loadbalancer for path optimization.
+2. Some LBs are sending packet with the destination IP equal to the LB IP, so the IP needs to be routable directly to the right service (otherwise looping back and forth)
+3. (TBD)
+
+This enhancement proposes a way to make this features configurable, giving the cloud provider, or the user, a way to disable this behaviour of kube proxy.
+The best way would be to set a defaut behaviour at the cloud controller level, that can be overwritten bu the user at a service level.
+
+## Motivation
+
+There are numerous problems with the current behaviour:
+
+1. Some cloud provider (Scaleway, ...) are sometimes using the LB's external IP as source IP when sending packets to the cluster. This is a probem in the ipvs mode of kube proxy since the IP is bouded to an interface and healtchecks from the LB are never coming back.
+2. Some cloud provider (DigitalOcean, Scaleway, ...) are also having features at the LB level (TLS termination, PROXY protocol, ...) and bypassing the LB means missing these features when the packet arrives to the service (leading to protocols errors).
+
+So giving an options to these cloud to disable the actual beahviour would be very valuable.
+
+Currently there is some hacky workaround that set the `Hostname` on the `Service` in order to bypass kube proxy binding (AWS and DigitalOcean for instance), but as stated, it's a bit hacky. This KEP should also give a solution to this hack.
+
+### Goals
+
+* Expose an option (ideally for at least the cloud controller manager) to disable the actual behaviour of kube proxy
+
+### Non-Goals
+
+* Deprecate the actual behaviour of kube proxy
+* Change the default actual behaviour of kube proxy
+
+## Proposal
+
+Right now several options are possible:
+
+* Add a flag to kube proxy to change the behaviour for all external IPs
+* Add a field to `LoadBalancerIngress` to disable the actual behaviour for a specific service
+* Add a new field to the `Service` object to disable the actual behaviour for a specific service
+
+I have the following proposal:
+
+- Add a `LoadBalancerIPBinding *bool` field to the `ServiceSpec` (will gladly change the name)
+- This variable will stay `nil` until the service is changed to a `LoadBalancer` type.
+- When the `Service` changes to LB, the cloud controller will query the cloud controller manager to get the default value (through a new method in the interface) if it's still `nil`.
+- The kube-proxy will be able to see the `LoadBalancerIPBinding` and act upon it on wheras the IP should be bound to the nodes or not.
```

Comment:
I would believe we want to flip the attribute right (say DisableIpBinding or something like that). So that way if there are no changes in to the cloud controller manager to react to LB IP allocation, then the default behavior becomes the current behavior. This would make it easier to rollout.

But having said that, since we are already using annotations for service load balancer behaviour, why wouldn't this also be an annotation added by the user? Or maybe a kube-proxy flag if it applies to all services in an environment?

Reply:
Yes, bad naming! The intention was to keep to actual behaviour the default one :) 

Annotations are also fine I guess. The kube proxy flag seems a bit too much, except if it can be overwritten by annotations though. 

As I said, it would be nice to have a default value (either given by a flag in kube proxy, or by a cloud controller method) and give the user an option to set his own wanted behaviour